src/Configuration/Programmer/Programmer_helper.cpp

```diff
@@ -0,0 +1,453 @@
+  std::string token;
+  size_t pos = -1;
+  std::vector<std::string> matches;
+  const std::string pattern(
+      R"(\s*(\d+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\d+)\s+(\S+)\s+(\S+)\s*)");
+  while (std::getline(iss, line)) {
+    matches = findStringPattern(line, pattern);
+    if (matches.size() == 0) {
+      continue;
+    }
+    // Replace all the spaces with a single space
+    while ((pos = line.find("  ")) != std::string::npos) {
+      line.replace(pos, 2, " ");
+    }
+    std::istringstream lineStream(line);
+    std::vector<std::string> tokens;
+    while (lineStream >> token) {
+      tokens.push_back(token);
+    }
+    if (tokens.size() == 8) {
+      TapInfo tapInfo;
+      tapInfo.index = CFG_convert_string_to_u64(tokens[0]);
+      tapInfo.tapName = tokens[1];
+      tapInfo.enabled = tokens[2] == "Y" ? true : false;
+      tapInfo.idCode =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[3], 0, 0));
+      tapInfo.expected =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[4], 0, 0));
+      tapInfo.irLen =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[5]));
+      tapInfo.irCap =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[6], 0, 0));
+      tapInfo.irMask =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[7], 0, 0));
+      tapInfoList.push_back(tapInfo);
+    }
+  }
+
+  return tapInfoList;
+}
+/*
+This function expecting the string from openOCD gemini list output
+with example below
+
+         Device               ID           IRLen      Flash Size
+-------- -------------------- ------------ ---------- ----------------
+Found  0 Gemini               0x20000913   5          16384
+*/
+std::vector<Device> extractDeviceList(const std::string& devicesString) {
+  std::vector<Device> devices;
+
+  std::istringstream iss(devicesString);
+  std::string line;
+  size_t pos = -1;
+  std::vector<std::string> matches;
+  const std::string pattern(
+      R"(Found\s+(\d+)\s+([\w\s]+)\s+([0-9a-fA-Fx]+)\s+(\d+)\s+(\d+))");
+
+  while (std::getline(iss, line)) {
+    matches = findStringPattern(line, pattern);
+    if (matches.size() == 0) {
+      continue;
+    }
+    // Replace all the spaces with a single space
+    while ((pos = line.find("  ")) != std::string::npos) {
+      line.replace(pos, 2, " ");
+    }
+    Device device;
+    std::istringstream lineStream(line);
+    std::string firstToken;
+    std::getline(lineStream, firstToken, ' ');  // "found string"
+
+    std::string indexStr;
+    std::getline(lineStream, indexStr, ' ');  // index string
+    device.index = static_cast<int>(CFG_convert_string_to_u64(indexStr));
+
+    std::getline(lineStream >> std::ws, device.name, ' ');  // device name
+
+    std::string idCodeStr;
+    std::getline(lineStream >> std::ws, idCodeStr,
+                 ' ');  // idCode string in hex
+    device.tapInfo.idCode =
+        static_cast<uint32_t>(CFG_convert_string_to_u64(idCodeStr));
```

Comment:
Is that expected we might lose data due to all conversions like this ?

Reply:
the idCode is unsigned 32bit. So the data lose doesn't matter.